Mathlib/Data/Polynomial/Smeval.lean

```diff
@@ -0,0 +1,210 @@
+
+variable (r : R) (x : S) (p q : R[X])
+
+theorem smeval_mul_X : (p * X).smeval x = p.smeval x * x := by
+    induction p using Polynomial.induction_on' with
+  | h_add p q ph qh =>
+    simp only [add_mul, smeval_add, ph, qh]
+  | h_monomial n a =>
+    rw [← monomial_one_one_eq_X, monomial_mul_monomial, smeval_monomial, smeval_monomial, mul_one,
+    pow_succ', smul_mul_assoc]
+
+theorem smeval_X_mul : (X * p).smeval x = x * p.smeval x := by
+    induction p using Polynomial.induction_on' with
+  | h_add p q ph qh =>
+    simp only [smeval_add, ph, qh, mul_add]
+  | h_monomial n a =>
+    rw [← monomial_one_one_eq_X, monomial_mul_monomial, smeval_monomial]
+    rw [one_mul, pow_add, pow_one, ← mul_smul_comm, smeval_monomial]
+
+theorem smeval_C_mul : (C r * p).smeval x = r • p.smeval x := by
+  induction p using Polynomial.induction_on' with
+  | h_add p q ph qh =>
+    simp only [mul_add, smeval_add, ph, qh, smul_add]
+  | h_monomial n b =>
+    simp only [C_mul_monomial, smeval_monomial, mul_smul]
+
+theorem smeval_mul_X_pow : ∀(n : ℕ), (p * X^n).smeval x = p.smeval x * x^n
+  | 0 => by
+    simp only [pow_zero, mul_one]
+  | n + 1 => by
+    rw [pow_add, ← mul_assoc, pow_one, smeval_mul_X, smeval_mul_X_pow n, pow_add, mul_assoc,
+      pow_one]
+
+theorem smeval_X_pow_mul : ∀(n : ℕ), (X^n * p).smeval x = x^n * p.smeval x
+  | 0 => by
+    simp [pow_zero, one_mul]
+  | n + 1 => by
+    rw [add_comm, pow_add, mul_assoc, pow_one, smeval_X_mul, smeval_X_pow_mul n, pow_add, mul_assoc,
+      pow_one]
+
+theorem smeval_monomial_mul (r : R) (n : ℕ) :
+    smeval x (monomial n r * p) = r • (x ^ n * p.smeval x) := by
+  induction p using Polynomial.induction_on' with
+  | h_add r s hr hs =>
+    simp only [add_comp, hr, hs, smeval_add, add_mul]
+    rw [← @C_mul_X_pow_eq_monomial, mul_assoc, smeval_C_mul, smeval_X_pow_mul, smeval_add]
+  | h_monomial n a =>
+    rw [smeval_monomial, monomial_mul_monomial, smeval_monomial, pow_add, mul_smul, mul_smul_comm]
+
+theorem smeval_mul : (p * q).smeval x  = p.smeval x * q.smeval x := by
+  induction p using Polynomial.induction_on' with
+  | h_add r s hr hs =>
+    simp only [add_comp, hr, hs, smeval_add, add_mul]
+  | h_monomial n a =>
+    simp only [smeval_monomial, smeval_C_mul, smeval_mul_X_pow, smeval_monomial_mul, smul_mul_assoc]
+
+end Algebra
+
+section Comparisons
+
+theorem eval₂_eq_smeval (R : Type*) [Semiring R] {S : Type*} [Semiring S] (f : R →+* S) (p : R[X])
+    (x: S) : haveI : Module R S := RingHom.toModule f
+    p.eval₂ f x = p.smeval x := by
+  letI : Module R S := RingHom.toModule f
+  rw [smeval_eq_sum, eval₂_eq_sum]
+  exact rfl
+
+theorem leval_eq_smeval {R : Type*} [Semiring R] (r : R) : leval r =
+    {
+      toFun := fun p => p.smeval r
+      map_add' := fun p q => smeval_add R p q r
+      map_smul' := by
+        intro r p
+        simp only [smeval_smul, smul_eq_mul, RingHom.id_apply]
+    } := by
+  aesop
```

Comment:
I think that for rewriting purposes, the lemma is more useful if you state it on the level of (coerced) functions. Because then you can also rewrite backwards, to turn a `smeval` into a `leval` when that is possible/useful.

Reply:
Thank you.  This would not have occurred to me.